src/main/java/org/elasticsearch/index/query/BaseQueryBuilder.java

```diff
@@ -22,17 +22,30 @@
 import org.apache.lucene.search.Query;
 import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentFactory;
 import org.elasticsearch.common.xcontent.XContentType;
 
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 /**
  * Base class with common code for all {@link QueryBuilder} implementations.
  */
 public abstract class BaseQueryBuilder implements QueryBuilder {
 
+    private final static Map<String, Class<?>> queryBuilderRegistry = new HashMap<>();
```

Comment:
If we merge parsers and builders at some point anyway why not re-use the registry for query parsers we already have in place and add a "get builder" method to the existing query parsers?

Something along the lines of:
    public abstract class BaseQueryBuilder implements QueryBuilder {
        IndexQueryParserService service;
    
        @Inject
        public BaseQueryBuilder(IndexQueryParserService service) {
            this.service = service;
        }

        protected QueryBuilder readQueryBuilderFrom(StreamInput in) throws IOException {
            QueryBuilder newBuilder;
            String name = in.readString();

            // the following two lines would become one once parsers and builders are merged
            QueryParser parser = this.service.queryParser(name);
            QueryBuilder builder = parser.createBuilder();
            builder.readFrom(in);
        }